README.md
A builder for Wave files.  This utility can create the byte list to write a wave file,
concatenating other wave files and periods of silence.

This utility is designed specifically to make metronome files (click tracks), but can be used for other applications as well.

Created from templates made available by Stagehand under a BSD-style
[license](https://github.com/dart-lang/stagehand/blob/master/LICENSE).

## Usage

A simple usage example:

```dart
import 'package:wave_builder/wave_builder.dart';

/// Create 1 bar of a 4/4 metronome at 60 BPM
main() {
  var fileOut = File('./example/assets/out/test.wav');
  var primary = File('./example/assets/wav/primary.wav');
  var secondary = File('./example/assets/wav/secondary.wav');
  var silenceType = WaveBuilderSilenceType.BeginningOfLastSample;

  var primaryBytes = await primary.readAsBytes();
  var secondaryBytes = await secondary.readAsBytes();

  await fileOut.create();

  var waveBuilder = WaveBuilder();

  waveBuilder.appendFileContents(primaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  waveBuilder.appendFileContents(secondaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  waveBuilder.appendFileContents(secondaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  waveBuilder.appendFileContents(secondaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  await fileOut.writeAsBytes(waveBuilder.fileBytes);
}
```


CHANGELOG.md
## 1.0.0

- Initial version, created by Stagehand
- Supports adding wave files and silence.
- Bit rate, frequency, and number of channels (mono/stereo) are hardcoded.

## 1.0.1

- Fixed a couple typos in docs, created/updated docs.
- Allow audio properties (frequency, bit rate, and number of channels) to be passed in instead of hardcoded.

## 1.0.2

- No code changes, but homepage is now reachable (so it will pass that analysis).

## 1.0.3

- Updated home page to github repository.

## 1.0.4

- Optimizations to adding files.
- Supports finding the data chunk first, to prevent repeating the iteration many times.

## 1.0.5

- Increase minimum SDK to allow on more recent dart versions (null safety)

test/wave_builder_test.dart
import 'package:wave_builder/src/lib/byte_utils.dart';
import 'package:wave_builder/wave_builder.dart';
import 'package:test/test.dart';

const int SAMPLE_SIZE = 4;
const int BIT_RATE = 16;
const int FREQUENCY = 44100;

const int RIFF_CHUNK_SIZE_INDEX = 4;
const int SUB_CHUNK_SIZE = 16;
const int AUDIO_FORMAT = 1;
const int BYTE_SIZE = 8;

// 2 for stereo
const int NUM_CHANNELS = 2;

void main() {
  group('WaveBuilder', () {
    var waveBuilder = WaveBuilder();
    setUp(() {
      waveBuilder = WaveBuilder();
    });

    group('#constructor', () {
      test('it initializes outputBytes and initializes chunks', () {
        expect(waveBuilder.fileBytes.length, 44);
      });

      test('it adds RIFF chunk', () {
        var riff = 'RIFF'.codeUnits;
        expect(waveBuilder.fileBytes.getRange(0, 4), riff);
        var wave = 'WAVE'.codeUnits;

        // do not test length here, because it is already calculated as non-zero
        // (file length minus this section)

        expect(waveBuilder.fileBytes.getRange(8, 12), wave);
      });

      test('it adds format chunk', () {
        var byteRate = FREQUENCY * NUM_CHANNELS * BIT_RATE ~/ BYTE_SIZE,
            blockAlign = NUM_CHANNELS * BIT_RATE ~/ 8,
            bitsPerSample = BIT_RATE;

        var fmt = 'fmt '.codeUnits;
        expect(waveBuilder.fileBytes.getRange(12, 16), fmt);
        expect(waveBuilder.fileBytes.getRange(16, 20),
            ByteUtils.numberAsByteList(SUB_CHUNK_SIZE, 4, bigEndian: false));

        expect(waveBuilder.fileBytes.getRange(20, 22),
            ByteUtils.numberAsByteList(AUDIO_FORMAT, 2, bigEndian: false));

        expect(waveBuilder.fileBytes.getRange(22, 24),
            ByteUtils.numberAsByteList(NUM_CHANNELS, 2, bigEndian: false));

        expect(waveBuilder.fileBytes.getRange(24, 28),
            ByteUtils.numberAsByteList(FREQUENCY, 4, bigEndian: false));

        expect(waveBuilder.fileBytes.getRange(28, 32),
            ByteUtils.numberAsByteList(byteRate, 4, bigEndian: false));

        expect(waveBuilder.fileBytes.getRange(32, 34),
            ByteUtils.numberAsByteList(blockAlign, 2, bigEndian: false));

        expect(waveBuilder.fileBytes.getRange(34, 36),
            ByteUtils.numberAsByteList(bitsPerSample, 2, bigEndian: false));
      });

      test('it adds data chunk', () {
        var data = 'data'.codeUnits;
        expect(waveBuilder.fileBytes.getRange(36, 40), data);
        expect(waveBuilder.fileBytes.getRange(40, 44), [0, 0, 0, 0]);
      });
    });

    group('#appendFileContents', () {
      test('it appends file contents after "data" string', () {
        var testNewFile = [1, 1, 1];
        testNewFile.addAll('data'.codeUnits);

        // size chunk
        testNewFile.addAll([0, 0, 0, 4]);

        testNewFile.addAll([1, 2, 3, 4]);
        waveBuilder.appendFileContents(testNewFile);
        expect(waveBuilder.fileBytes.getRange(44, 48), [1, 2, 3, 4]);
      });

      test('it appends all file contents if data string does not exist', () {
        var testNewFile = [1, 1, 1, 2];
        waveBuilder.appendFileContents(testNewFile);
        expect(waveBuilder.fileBytes.getRange(44, 48), [1, 1, 1, 2]);
      });
    });

    group('#appendSilence', () {
      group('End of last sample', () {
        test('it adds 0s for appropriate length (when appended to end)', () {
          var testNewFile = [1, 1, 1, 2];
          var byteRate = FREQUENCY * BIT_RATE ~/ 8;
          var silenceLength = byteRate * NUM_CHANNELS;
          waveBuilder.appendFileContents(testNewFile);
          waveBuilder.appendSilence(
              1000, WaveBuilderSilenceType.EndOfLastSample);
          expect(waveBuilder.fileBytes.getRange(44, 48), [1, 1, 1, 2]);
          expect(waveBuilder.fileBytes[49], 0);
          expect(waveBuilder.fileBytes[48 + silenceLength - 1], 0);
          expect(waveBuilder.fileBytes.length, 48 + silenceLength);
        });
      });

      group('Beginning of last sample', () {
        test('it adds 0s for length minus last sample size', () {
          var testNewFile = [1, 1, 1, 2];
          var byteRate = FREQUENCY * BIT_RATE ~/ 8;
          var silenceLength = byteRate * NUM_CHANNELS;
          waveBuilder.appendFileContents(testNewFile);
          waveBuilder.appendSilence(
              1000, WaveBuilderSilenceType.BeginningOfLastSample);
          expect(waveBuilder.fileBytes.getRange(44, 48), [1, 1, 1, 2]);
          expect(waveBuilder.fileBytes[49], 0);
          expect(waveBuilder.fileBytes[48 + silenceLength - 4 - 1], 0);
          expect(waveBuilder.fileBytes.length, 48 + silenceLength - 4);
        });
        test(
            'it truncates last sample appropriately if silence is less than last sample size',
            () {
          var byteRate = FREQUENCY * BIT_RATE ~/ 8;
          var silenceLength = byteRate * NUM_CHANNELS;
          var testNewFile = List<int>.filled(silenceLength + 4, 1);
          waveBuilder.appendFileContents(testNewFile);
          waveBuilder.appendSilence(
              1000, WaveBuilderSilenceType.BeginningOfLastSample);
          expect(waveBuilder.fileBytes.length, 44 + silenceLength);
          expect(waveBuilder.fileBytes[48 + 1], 1);
          expect(waveBuilder.fileBytes[testNewFile.length - 1], 1);
        });
      });
    });

    group('#finalize', () {
      test('it updates riff chunk size', () {
        // finalize is called for every fileBytes getter
        waveBuilder.appendFileContents(<int>[1, 2, 3, 4]);
        expect(waveBuilder.fileBytes[RIFF_CHUNK_SIZE_INDEX], 40);
      });

      test('it updates data chunk size', () {
        waveBuilder.appendFileContents(<int>[1, 2, 3, 4]);
        expect(waveBuilder.fileBytes[40], 4);
      });
    });

    group('#findDataChunk', () {
      test('it finds the data chunk (list contents after "data" + size)', () {
        var testNewFile = [1, 1, 1];
        testNewFile.addAll('data'.codeUnits);

        // size chunk
        testNewFile.addAll([0, 0, 0, 4]);

        testNewFile.addAll([1, 2, 3, 4]);

        expect(waveBuilder.getDataChunk(testNewFile), [1, 2, 3, 4]);
      });
    });
  });
}


test/lib/byte_utils_test.dart
import 'package:test/test.dart';
import 'package:wave_builder/src/lib/byte_utils.dart';

void main() {
  group('ByteUtils', () {
    group('#numberAsByteList', () {
      test('it returns 2 byte number properly', () {
        var testVal = 258;
        expect(ByteUtils.numberAsByteList(testVal, 2), [1, 2]);
      });

      test('it returns padded 4 byte number properly', () {
        var testVal = 65538;
        expect(ByteUtils.numberAsByteList(testVal, 4), [0, 1, 0, 2]);
      });

      test('it works for little endian', () {
        var testVal = 65538;
        expect(ByteUtils.numberAsByteList(testVal, 4, bigEndian: false),
            [2, 0, 1, 0]);
      });
    });

    group('#findByteSequenceInList', () {
      test('it finds a sequence in a list and returns the index', () {
        var list = <int>[5, 4, 3, 8, 8, 6, 7, 5, 3, 0, 9, 8];
        var sequence = <int>[8, 6, 7, 5, 3, 0, 9];
        expect(ByteUtils.findByteSequenceInList(sequence, list), 4);
      });

      test(
          'it returns -1 if sequence not found (does not err on index overflow)',
          () {
        var list = <int>[5, 4, 3, 8, 8, 6, 7, 5, 3, 0, 9, 8];
        var sequence = <int>[8, 6, 7, 5, 3, 0, 8];
        expect(ByteUtils.findByteSequenceInList(sequence, list), -1);
      });
    });
  });
}


example/wave_builder_example.dart
import 'dart:io';

import 'package:wave_builder/wave_builder.dart';

Future<void> main() async {
  var fileOut = File('./example/assets/out/test.wav');
  var primary = File('./example/assets/wav/primary.wav');
  var secondary = File('./example/assets/wav/secondary.wav');
  var silenceType = WaveBuilderSilenceType.BeginningOfLastSample;

  var primaryBytes = await primary.readAsBytes();
  var secondaryBytes = await secondary.readAsBytes();

  if (await fileOut.exists()) {
    await fileOut.delete();
  }

  await fileOut.create();

  var waveBuilder = WaveBuilder();

  waveBuilder.appendFileContents(primaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  waveBuilder.appendFileContents(secondaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  waveBuilder.appendFileContents(secondaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  waveBuilder.appendFileContents(secondaryBytes);
  waveBuilder.appendSilence(1000, silenceType);
  await fileOut.writeAsBytes(waveBuilder.fileBytes);
}


lib/src/wave_builder_base.dart
import 'dart:convert';

import 'package:wave_builder/src/lib/byte_utils.dart';

/// This flag is used when appending silence.
/// Determines whether or not to start the silence timer at the beginning or end of the
/// last sample appended to this file.
enum WaveBuilderSilenceType { BeginningOfLastSample, EndOfLastSample }

/// Build a wave file.
class WaveBuilder {
  static const int RIFF_CHUNK_SIZE_INDEX = 4;
  static const int SUB_CHUNK_SIZE = 16;
  static const int AUDIO_FORMAT = 1;
  static const int BYTE_SIZE = 8;

  int _lastSampleSize = 0;

  /// Finalizes the header sizes and returns bytes
  List<int> get fileBytes {
    _finalize();
    return _outputBytes;
  }

  List<int> _outputBytes = <int>[];
  final Utf8Encoder _utf8encoder = Utf8Encoder();

  int _dataChunkSizeIndex = 0;

  int _bitRate = 0;
  int _frequency = 0;
  int _numChannels = 0;

  /// Construct a wave builder.
  /// Supply audio file properties.
  WaveBuilder({int bitRate = 16, int frequency = 44100, bool stereo = true}) {
    _outputBytes = <int>[];
    _bitRate = bitRate;
    _frequency = frequency;
    _numChannels = stereo ? 2 : 1;
    _initializeWave();
  }

  void _initializeWave() {
    _outputBytes.addAll(_utf8encoder.convert('RIFF'));
    _outputBytes.addAll(ByteUtils.numberAsByteList(0, 4, bigEndian: false));
    _outputBytes.addAll(_utf8encoder.convert('WAVE'));

    _createFormatChunk();
    _writeDataChunkHeader();
  }

  void _createFormatChunk() {
    var byteRate = _frequency * _numChannels * _bitRate ~/ BYTE_SIZE,
        blockAlign = _numChannels * _bitRate ~/ 8,
        bitsPerSample = _bitRate;
    _outputBytes.addAll(_utf8encoder.convert('fmt '));
    _outputBytes.addAll(
        ByteUtils.numberAsByteList(SUB_CHUNK_SIZE, 4, bigEndian: false));
    _outputBytes
        .addAll(ByteUtils.numberAsByteList(AUDIO_FORMAT, 2, bigEndian: false));
    _outputBytes
        .addAll(ByteUtils.numberAsByteList(_numChannels, 2, bigEndian: false));
    _outputBytes
        .addAll(ByteUtils.numberAsByteList(_frequency, 4, bigEndian: false));
    _outputBytes
        .addAll(ByteUtils.numberAsByteList(byteRate, 4, bigEndian: false));
    _outputBytes
        .addAll(ByteUtils.numberAsByteList(blockAlign, 2, bigEndian: false));
    _outputBytes
        .addAll(ByteUtils.numberAsByteList(bitsPerSample, 2, bigEndian: false));
  }

  void _writeDataChunkHeader() {
    _outputBytes.addAll(_utf8encoder.convert('data'));
    _dataChunkSizeIndex = _outputBytes.length;
    _outputBytes.addAll(ByteUtils.numberAsByteList(0, 4, bigEndian: false));
  }

  /// Find data chunk content after <data|size> in [fileContents]
  List<int> getDataChunk(List<int> fileContents) {
    final dataIdSequence = _utf8encoder.convert('data');
    final dataIdIndex =
        ByteUtils.findByteSequenceInList(dataIdSequence, fileContents);
    var dataStartIndex = 0;

    if (dataIdIndex != -1) {
      // Add 4 for data size
      dataStartIndex = dataIdIndex + dataIdSequence.length + 4;
    }
    return fileContents.sublist(dataStartIndex);
  }

  /// Append [fileContents] read as bytes to our wave file.
  /// If [findDataChunk] is true, searches first to find the file's data chunk.
  /// It's recommended to call getDataChunk on the file contents you want to append first,
  /// to prevent repeating everytime you add the same file.
  void appendFileContents(List<int> fileContents, {bool findDataChunk = true}) {
    var dataChunk = findDataChunk ? getDataChunk(fileContents) : fileContents;
    _lastSampleSize = dataChunk.length;
    _outputBytes.addAll(dataChunk);
  }

  /// Append [msLength] milliseconds of silence to our wave file.
  /// [silenceType] determines whether we start the counter for silence
  /// from the beginning of the last sample or the end.
  void appendSilence(int msLength, WaveBuilderSilenceType silenceType) {
    var byteRate = _frequency * _bitRate ~/ 8;
    var length = (msLength * byteRate ~/ 1000) * _numChannels;
    if (silenceType == WaveBuilderSilenceType.BeginningOfLastSample) {
      if (length > _lastSampleSize) {
        length -= _lastSampleSize;
      } else {
        _outputBytes.removeRange(_outputBytes.length - _lastSampleSize + length,
            _outputBytes.length);
        length = 0;
      }
    }

    _outputBytes.addAll(List<int>.filled(length, 0));
  }

  void _updateRiffChunkSize() {
    _outputBytes.replaceRange(
        RIFF_CHUNK_SIZE_INDEX,
        RIFF_CHUNK_SIZE_INDEX + 4,
        ByteUtils.numberAsByteList(
            _outputBytes.length - (RIFF_CHUNK_SIZE_INDEX + 4), 4,
            bigEndian: false));
  }

  void _updateDataChunkSize() {
    _outputBytes.replaceRange(
        _dataChunkSizeIndex,
        _dataChunkSizeIndex + 4,
        ByteUtils.numberAsByteList(
            _outputBytes.length - (_dataChunkSizeIndex + 4), 4,
            bigEndian: false));
  }

  void _finalize() {
    _updateRiffChunkSize();
    _updateDataChunkSize();
  }
}
